tests/Form/LayoutTest.php

```diff
@@ -0,0 +1,64 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Atk4\Ui\Tests\Form;
+
+use Atk4\Core\Phpunit\TestCase;
+use Atk4\Data\Model;
+use Atk4\Ui\Button;
+use Atk4\Ui\Form;
+use Atk4\Ui\HtmlTemplate;
+use Atk4\Ui\Tests\CreateAppTrait;
+use Atk4\Ui\ViewWithContent;
+
+class LayoutTest extends TestCase
+{
+    use CreateAppTrait;
+
+    public function testRecursiveRenderRespectsRegions(): void
+    {
+        $template = new HtmlTemplate(
+            '<div id="defaultRegion">{$Content}{$Buttons}</div><div id="anotherRegion">{$input2}{$AnotherRegion}</div>'
+        );
+        $form = new Form(
+            [
+                'buttonSave' => false,
+                'layout' => [
+                    Form\Layout::class,
+                    'template' => $template,
+                ],
+            ]
+        );
+        $form->setApp($this->createApp());
+        $form->invokeInit();
+
+        $m = new Model();
+        $m->addField('input1');
+        $m->addField('input2');
+        $form->setEntity($m->createEntity());
+
+        // all of these should be rendered inside #defaultRegion
+        Button::addTo($form->layout, ['Button1']);
+        ViewWithContent::addTo($form->layout, ['View1']);
+        Form\Layout::addTo($form->layout, ['label' => 'Layout1']);
+
+        // all of these should be rendered inside #anotherRegion
+        Button::addTo($form->layout, ['Button2'], ['AnotherRegion']);
+        ViewWithContent::addTo($form->layout, ['View2'], ['AnotherRegion']);
+        Form\Layout::addTo($form->layout, ['label' => 'Layout2'], ['AnotherRegion']);
+
+        $renderedHtml = $form->getHtml();
+        $firstPart = substr($renderedHtml, 0, strpos($renderedHtml, 'id="anotherRegion"'));
```

Comment:
This is too fragile and hard to read. Why not to use something like contains string and assert the whole inner html 1:1?

Reply:
asserting the whole HTML now